### PR TITLE
Implement exercise micro service

### PR DESCRIPTION
**Pull request by Dieuer**
## Description
- Implements the exercise micro service. It interacts with the MongoDB and provides endpoints to retrieve a specific exercise by id or all exercises. These two endpoints supports the get exercise and get exercises use cases. The micro service makes use of the template created by Andreas and Freja. Extended from the template we also include an .env configuration file and altered CI and Dockerfile which works without a package lock file. 

### Type of change
- [x] New feature (non-breaking change which adds functionality)

### How Has This Been Tested?
*Please describe the tests that you ran to verify your changes*

### Checklist:

- [x] You have added a clear title and description for the PR
- [x] The CI pass
- [x] There are no merge conflicts
- [x] The code satisfies the [code standards](https://docs.google.com/document/d/1R9n9MeuRwz3WSrnWZUjrbmGeosTlu1Xz8f5F7gu2AJE/edit#heading=h.e9xl4xfoldow) 
- [x] I have added docstrings for my code, where needed
- [x] Code implements intended functionality
- [x] All relevant tests pass (Black, White, Integration, Monkey, etc.)

### CODE REVIEW

File: src/application/actors/index.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './user';
```
Comment (JakobFaarGreg) [inline src/application/actors/index.ts:1]: What is the reason for doing this?

File: src/application/actors/user.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    request: IGetExercisesRequest,
+  ): Promise<IGetExercisesResponse> {
+    return this.getExercisesUseCase.do(request);
+  }
+
+  public async getExercise(
+    request: IGetExerciseRequest,
+  ): Promise<IGetExerciseResponse> {
+    return this.getExerciseUseCase.do(request);
+  }
```
Comment (JakobFaarGreg) [inline src/application/actors/user.ts]: Er der en mulighed for at de her to 'getExerciseUseCase' og  'getExercisesUseCase' på en måde der gør dem lettere genkendelige

File: src/application/usecases/get_exercises_use_case.spec.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { Exercise, IExerciseRepository } from '../../domain';
+import { mock } from 'jest-mock-extended';
+import { GetExercisesUseCase } from './get_exercises_use_case';
+import { v4 as uuidv4 } from 'uuid';
+
+describe('do', () => {
+  it('should return an empty list of exercises if non were found', async () => {
```
Comment (JakobFaarGreg) [inline src/application/usecases/get_exercises_use_case.spec.ts]:
```suggestion
  it('should return an empty list of exercises if none were found', async () => {
```

File: src/application/usecases/get_exercises_use_case.spec.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    // Mock
+    repository.getExercises.mockResolvedValue([]);
+
+    // Act
+    const actual = await exercisesUseCase.do({});
+
+    // Assert
+    expect(actual.exercises).toEqual([]);
+  }),
+    it('should return an empty list of exercises if non were found', async () => {
```
Comment (JakobFaarGreg) [inline src/application/usecases/get_exercises_use_case.spec.ts]:
```suggestion
    it('should return an empty list of exercises if none were found', async () => {
```

File: src/application/usecases/get_exercises_use_case.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+export class GetExercisesUseCase
+  implements IUseCase<IGetExercisesRequest, IGetExercisesResponse>
+{
+  private readonly repository: IExerciseRepository;
+
+  constructor(repository: IExerciseRepository) {
+    this.repository = repository;
+  }
+
+  public async do(_: IGetExercisesRequest): Promise<IGetExercisesResponse> {
```
Comment (JakobFaarGreg) [inline src/application/usecases/get_exercises_use_case.ts]: Ligesom den Github Action siger, hvad har vi brug for parametret til? Hvis vi bare kalder det '_'

File: src/domain/exercise.spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import { Exercise } from './exercise';
+import { Query } from './query';
+
+describe('Constructor', () => {
+  it('throws an error if the id is not a uuidv4', () => {
+    expect(() => new Exercise('asd', 'a', 'a', [])).toThrow(Error);
+  }),
+    it('Throws an error if the title string is empty', () => {
+      // Assert
+      expect(() => new Exercise(uuidv4(), '', 'a', [])).toThrow(Error);
```
Comment (JakobFaarGreg) [inline src/domain/exercise.spec.ts]: This indentation seems weird, can we solve it?

File: src/domain/exercise.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { Query } from './query';
+
+function isUuidv4(input: string): boolean {
+  const regexExp =
+    /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/gi;
```
Comment (JakobFaarGreg) [inline src/domain/exercise.ts:5]:
```suggestion
    /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;
```
Vi burde ikke have brug for hverken global match eller case insensitive. Når REGEX er skrevet som det er.
ellers kunne det være: `/^[0-9A-F]{8}\b-[0-9A-F]{4}\b-[0-9A-F]{4}\b-[0-9A-F]{4}\b-[0-9A-F]{12}$/i`

File: src/domain/query.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  // Safety property: "E[] p"
+  if (input.startsWith('E[] ')) {
+    return true;
+  }
+
+  // Liveness property: "A<> p"
+  if (input.startsWith('A<> ')) {
+    return true;
+  }
```
Comment (JakobFaarGreg) [inline src/domain/query.ts:26]: Er det her en oplagt kandidat til at være en switch? Den kunne være nem at extende

File: src/index.ts [hunk 9]
```diff
@@ -1,6 +1,60 @@
-// eslint-disable-next-line no-console
-export function sum(a: number, b: number) {
-  return a + b;
+import fastify from 'fastify';
+import { MongoExerciseRepository } from './infrastructure';
+import { exerciseController, statusController } from './presentation';
+import * as dotenv from 'dotenv';
+
+// Load the ".env" fiel from the root. Afterwards check all required environment bindings
```
Comment (JakobFaarGreg) [inline src/index.ts]:
```suggestion
// Load the ".env" file from the root. Afterwards check all required environment bindings
```

File: src/infrastructure/mongo_exercise_repository.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * A concrete implementation of the exercise repository interface
+ */
+export class MongoExerciseRepository implements IExerciseRepository {
+  private readonly exerciseSchmea: Schema<IExerciseDocument>;
+  private readonly exerciseModel: Model<IExerciseDocument>;
+
+  /**
+   * Construct a MongoDB repository from which we query exercises
+   * @param connection The conenction string to MongoDB
```
Comment (JakobFaarGreg) [inline src/infrastructure/mongo_exercise_repository.ts]:
```suggestion
   * @param connection The connection string to MongoDB
```

File: src/infrastructure/mongo_exercise_repository.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  ) {
+    // Create the schema with the document rules
+    this.exerciseSchmea = new Schema<IExerciseDocument>({
+      id: { type: String, required: true },
+      title: { type: String, required: true },
+      description: { type: String, required: true },
+      queries: {},
+    });
+
+    // The model adhereing to the document schema
```
Comment (JakobFaarGreg) [inline src/infrastructure/mongo_exercise_repository.ts]:
```suggestion
    // The model adhering to the document schema
```

File: src/infrastructure/mongo_exercise_repository.ts [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  private documentsToExercises(models: IExerciseDocument[]): Exercise[] {
+    const exercises = [];
+    for (const model of models) {
+      exercises.push(this.documentToExercise(model));
+    }
+    return exercises;
+  }
+
+  /**
+   * @returns All exercises which are stored persistently in MonogDB
```
Comment (JakobFaarGreg) [inline src/infrastructure/mongo_exercise_repository.ts]:
```suggestion
   * @returns All exercises which are stored persistently in MongoDB
```

File: src/presentation/exercise_controller.spec.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+import { IGetExerciseResponse, IGetExercisesResponse } from '../application';
+
+describe('/exercises', () => {
+  const server = fastify();
+  const repository = mock<IExerciseRepository>();
+  server.register(exerciseController, {
+    exerciseRepository: repository,
+  });
+
+  it('Returns status 200 if we successfully get exercises', async () => {
```
Comment (JakobFaarGreg) [inline src/presentation/exercise_controller.spec.ts:18]: Nogle steder har i stort forbogstav i jeres tests, andre steder ikke. Skal vi ikke lige vælge en?

**Review by JakobFaarGreg — CHANGES_REQUESTED** [review]
With these changes it looks good to me 👍

File: src/application/actors/index.ts (hunk 1 above)
Comment (Dieuer) [reply src/application/actors/index.ts:1]: It exports everything from the "user" file, such that when importing "actors" module we get everything from "user".

File: src/application/usecases/get_exercises_use_case.ts (hunk 5 above)
Comment (Dieuer) [reply src/application/usecases/get_exercises_use_case.ts]: Vi skal opfylde interfacet, men vi har ikke brug for den aktuelle parameter.

File: src/domain/query.ts (hunk 8 above)
Comment (Dieuer) [reply src/domain/query.ts:26]: Man kan ikke lave en switch på det.

**Review by fatimaosman99 — APPROVED** [review]

File: src/presentation/exercise_controller.spec.ts (hunk 13 above)
Comment (Dieuer) [reply src/presentation/exercise_controller.spec.ts:18]: Det er blevet til stort nu

File: src/application/actors/user.ts (hunk 2 above)
Comment (Dieuer) [reply src/application/actors/user.ts]: renamed til get_all_exercises og get_an_exercise

**Review by JakobFaarGreg — APPROVED** [review]
Great work.

**Pull request merged**